horace_core/sqw/@Experiment/Experiment.m

```diff
@@ -429,6 +489,84 @@
             std_form = check_sample_or_inst_array_and_return_std_form_(...
                 sample_or_instrument,class_base);
         end
+        
+        function obj = add_input_with_checks(obj, val,type)
+            % ADD_INPUT_WITH_CHECKS
+            % Add val to one of the unique_reference_container properties.
+            % This makes sure that the right thing is added and any
+            % duplicates are handled.
+            %
+            % Input
+            % -----
+            % - obj:   the Experiment object to which val is being added
+            % - val:   the object or container of objects which is being
+            %          added. this could be a unique_references_container
+            %          or unique_objects_container with baseclass 'type'
+            %          or an object which 'isa(,type)'
+            % - type:  string naming the baseclass to which val should
+            %          belong
+            %
+            % Output
+            % ------
+            % - obj:   the modified Experiment object where the value of
+            %          the field named by the calling form of val (using
+            %          inputname(val#argnum) is modified by adding val if
+            %          val is not a unique_references-container or setting
+            %          it if it is a unique_references_container
+
+            % get the name of the calling variable which enters as 'val'
+            field = [inputname(2) '_'];
+
+            if isa(val, 'unique_references_container') && ...
+                   strcmp(val.stored_baseclass,type)
+                % if size is right, overwrite default compressed container
+                if val.n_runs == obj.n_runs
+                    obj.(field) = val;
+                else
+                    error('HORACE:Experiment:invalid_argument', ...
+                          'input %d size must match number of runs',obj.n_runs);
+                end
+
+
+            elseif ( isa(val, type) &&                       ...
+                     numel(val) == obj.n_runs )              ...
+                   ||                                        ...                   
+                   ( isa(val, 'unique_objects_container') && ...
+                     val.n_runs == obj.n_runs             && ...
+                     strcmp(val.baseclass, type) )           ...
+                   ||                                        ...
+                   ( iscell(val)                          && ...
+                     numel(val) == obj.n_runs             && ...
+                     isa(val{1}, type) )
+
+                % add to default compressed container
+                obj.(field) = obj.(field).add(val);
+
+            elseif ( isa(val, type) &&                       ...
+                     numel(val) == 1 )                       ...
```

Comment:
Can use
```suggestion
                     isscalar(val) )                       ...
```

Reply:
I am nervous of multiple routes into numel with these unique containers and will keep the original as an identifier for any issues arising